File: usaco/silver/closing.cpp
#include <bits/stdc++.h>
using namespace std;

int N, M;
unordered_set<int> open;
vector<vector<int>> graph;

bool dfs(int u, unordered_set<int> &o) {
    o.erase(u);
    for (int v : graph[u])
        if (o.count(v))
            dfs(v, o);
    return o.empty();
}

int main() {
    freopen("closing.in", "r", stdin);
    freopen("closing.out", "w", stdout);
    cin >> N >> M;
    graph.resize(N+1);
    for (int i = 0; i < M; ++i) {
        int a, b; cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }
    for (int i = 1; i <= N; ++i)
        open.insert(i);
    for (int i = 0; i < N; ++i) {
        auto o = open;
        cout << (dfs(*begin(o), o) ? "YES" : "NO") << endl;
        int a; cin >> a;
        open.erase(a);
    }
}

File: usaco/silver/closing.py
# tle cuz python bad
import sys
from collections import defaultdict
intput = lambda : map(int, input().split())

def dfs(u, s):
    s.remove(u)
    for v in graph[u]:
        if v in s:
            dfs(v, s)
    return len(s) == 0

if __name__ == '__main__':
    sys.stdin  = open('closing.in',  'r')
    sys.stdout = open('closing.out', 'w')
    N, M = intput()
    graph = defaultdict(list)
    for _ in range(M):
        a, b = intput()
        graph[a].append(b)
        graph[b].append(a)
    open = set(range(1,N+1))
    for i in range(N):
        for e in open: break  # lol
        print("YES" if dfs(e, set(open)) else "NO")
        open.remove(int(input()))
